### PR TITLE
Updated new replicator to BLIP 3 protocol (INCOMPATIBLE)

### LINKED ISSUE #3136: Inter-document compression in BLIP replicator
The BLIP-based replication protocol sends each revision in a separate message. The messages are gzip-compressed, but because the compressor starts over with each document, we don't get as optimal compression as we did in 1.x where _bulk_docs sent a whole series of documents in one gzipped HTTP message. (The benefits are greatest for small documents.)

The BLIP protocol is being updated to version 3 to support inter-message compression. This is implemented in both BLIP-Cpp and go-blip. SG will need a few changes to adapt to changes in the Go API.

At the same time we should take out the 'unsupported' flag and make the replicator enabled by default.

### PR DESCRIPTION
**Pull request by snej**
This is an INCOMPATIBLE REPLICATOR CHANGE, not interoperable with Couchbase Lite 2 until the LiteCore `feature/compression_297` branch is merged (couchbase/couchbase-lite-core#297)

* New replicator is always enabled (no more 'replicator_2' config.)
* Replicator uses BLIP 3 protocol with inter-message compression.
* Compression level can be set with 'replicator_compression' config option (range is 0-9)

Fixes #3136
  
Before merging:

- [ ] Merge couchbase/couchbase-lite-core#328
- [ ] Merge cblite-ios / android / .net

### CODE REVIEW

**Comment by tleyden** [comment]
Related couchbase-lite-core change: https://github.com/couchbase/couchbase-lite-core/issues/328

**Comment by tleyden** [comment]
@snej I'm trying to do an end-to-end manual smoke-test, and I can't get past this error on Sync Gateway when cblite tries to connect:

```
2017-12-19T15:20:53.088-08:00 BLIP: Error: Client doesn't support WS protocol BLIP_3a2, only 
```

There is nothing after the "only", because cblite doesn't seem to be sending anything in the `Sec-WS-Protocols` headers.  

I'm double-checking my environment to make sure that I've built the correction version of litecore.  I'm on:

* couchbase-lite-ios: feature/2.0 branch
* litecore: feature/compression_297 branch 
* blip-cpp: feature/blip3

Anything besides the `Sec-WS-Protocols` header I can use to ensure that I'm using the correct version of litecore?  Here's my todo app console output: 

https://gist.github.com/tleyden/a4e3f48210c7e3d0f8c27fede1d7e799

and here are the websocket headers during the handshake:

screen shot 2017-12-19 at 3 35 23 pm

Note the missing `Sec-WS-Protocols` headers.

File: rest/blip_sync.go
Comment (tleyden) [inline rest/blip_sync.go]: If `conn.Close()` returns an error, it's being ignored.  It should log a warning here.

File: rest/blip_sync.go
Comment (tleyden) [inline rest/blip_sync.go]: Is there a companion line where it logs the fact that a `BLIP+WebSocket` connection was opened?

@adamcfraser any thoughts on whether the high level `BLIP+WebSocket` creation teardown logs belong in `HTTP` vs `HTTP+`?  It feels like putting in `HTTP` would make debugging a bit easier .. and if these connections are fairly long-lived, then the log volume increase might be acceptable.

File: rest/config.go
Comment (tleyden) [inline rest/config.go:90]: Can you add the default compression level to the comment?

Comment (tleyden) [reply rest/config.go:90]: @jamiltz what's the best process for adding docs for new params?  Github PR or Jira issue?

**Review by tleyden — CHANGES_REQUESTED** [review]
See inline comments

File: rest/config.go
Comment (jamesnocentini) [reply rest/config.go:90]: GitHub PR, here's one with a placeholder https://github.com/couchbaselabs/couchbase-mobile-portal/pull/837

File: rest/blip_sync.go
Comment (adamcfraser) [reply rest/blip_sync.go]: For consistency I think we should be using HTTP+ (that's used elsewhere for closed connections).

**Comment by tleyden** [comment]
@snej should Sync Gateway have the [proposeChanges](https://github.com/couchbase/couchbase-lite-core/wiki/Replication-Protocol#proposechanges) endpoint enabled?  @adamcfraser mentioned that LiteCore might not be supporting this yet.

File: rest/blip_sync.go
Comment (snej) [reply rest/blip_sync.go]: Line 87 is the companion line. This was directly copied from the WebSocket code in changes_api.go.

File: rest/config.go
Comment (snej) [reply rest/config.go:90]: I could, but it would tend to become obsolete if we ever change the default level in go-blip, especially since that's in a separate repo...

**Comment by snej** [comment]
@tleyden This PR adds proposeChanges. LiteCore should only be calling proposeChanges if it called changes and got an error. I'll take a look at that other issue soon.

**Comment by tleyden** [comment]
@snej this PR looks good to me.  Any objection if I merge this first and then the LiteCore changes can follow?

**Comment by snej** [comment]
**Don't merge this PR yet!** I'm doing some testing with the latest LiteCore and finding some errors -- more of the same UnexpectedEOF crap from the Go 'flate' package, but for some reason on small frames (~200 bytes).

**Comment by tleyden** [comment]
I rebased this branch off the latest master and did a force push.  This isn't strictly necessary, but I think this just makes the commit history a bit easier to reason about.

**Review by tleyden — APPROVED** [review]

**Comment by tleyden** [comment]
@snej  This is ready to merge once you've merged in the LiteCore + platform changes.  

The latest go-blip PR is in, and this SG branch is updated to point to the latest go-blip.

Also I've re-run manually smoke tests using the latest commit on this branch and the todo ios sample app on the `feature/2.0` branch.

**Pull request merged**
